### DISCUSSION TITLE
openapi allOf format

### DISCUSSION THREAD

**Discussion by tbruyelle** [cuelang/cue#940] [Q&A] [answered]
Hi cue team,

I have an issue with the openapi format generated by `cue export --out openapi+yaml` (version 0.3.2) with the following cue file :

```
#Response: {
	id:      uint
	result: {...}
}

#VersionResponse: #Response & {
	result: version: string
}
```

The output is :

```yaml
openapi: 3.0.0
info:
  title: Generated by cue.
  version: no version
paths: {}
components:
  schemas:
    Response:
      type: object
      required:
      - id
      - result
      properties:
        id:
          type: integer
          minimum: 0
        result:
          type: object
    VersionResponse:
      type: object
      properties:
        result:
          type: object
          required:
          - version
          properties:
            version:
              type: string
      allOf:
      - $ref: '#/components/schemas/Response'
      - required:
        - result
```

The problem comes from the `VersionResponse` schema, the properties seem split between the root of the schema and the `allOf` field.

Instead of 

```yaml
    VersionResponse:
      type: object
      properties:
        result:
          type: object
          required:
          - version
          properties:
            version:
              type: string
      allOf:
      - $ref: '#/components/schemas/Response'
      - required:
        - result
```

I think it should look like that with everything in the `allOf` field:

```yaml
    VersionResponse:
      allOf:
      - $ref: '#/components/schemas/Response'
      - type: object
        required:
          - result
        properties:
          result:
            type: object
            required:
            - version
            properties:
              version:
                type: string
```

WDYT?

**Comment by tbruyelle** [comment]
I noticed today that changing the conjunction to a disjunction, the resulting `oneOf` looks more correct as it contains every fields : 

 ```cue
#Response: {
	id:      uint
	result: {...}
}

#VersionResponse: #Response | {
	result: version: string
}
```

```yaml
 VersionResponse:
      oneOf:
        - $ref: '#/components/schemas/Response'
        - properties:
            result:
              properties:
                version:
                  type: string
              required:
                - version
              type: object
          required:
            - result
      type: object
```

**Comment by myitcv** [comment, marked as answer]
@tbruyelle (apologies for the delay in replying) - is there a difference in semantics? Or is this purely an aesthetic point aimed at improving readability (which would potentially be worthy in its own right)?

**Reply by tbruyelle** [reply]
@myitcv Well, at first glance I though this was an incorrect openapi format, because the library[1] I was using to parse it doesn't accept it. The library accepts the format with everything in `allOf`.

But actually it appears it's correct since it's totally accepted in the different renderers I tested (swagger-ui, redoc,...)

So I think I can close this discussion as non related ^^.

[1] https://github.com/getkin/kin-openapi

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/940.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
